src/ripple/nodestore/impl/DeterministicShard.cpp

```diff
@@ -0,0 +1,303 @@
+    {
+        remove_all(tempdir_);
+        preexist = false;
+    }
+
+    backend_->open(
+        !preexist,
+        digest(2) | 0x5348524400000000ll, /* appType */
+        digest(0),                        /* uid */
+        digest(1)                         /* salt */
+    );
+
+#ifndef NUDB_SUPPORTS_CUSTOM_UID
+    backend_->close();
+
+    {
+        boost::filesystem::path keypath = tempdir_ / "nudb.key";
+        boost::filesystem::path datpath = tempdir_ / "nudb.dat";
+
+        class stream_ : public std::fstream
+        {
+            std::streampos size_;
+
+        public:
+            stream_(const std::string& s)
+                : std::fstream(
+                      s,
+                      std::ios_base::in | std::ios_base::out |
+                          std::ios_base::binary)
+                , size_(0)
+            {
+            }
+            void
+            read(
+                size_t pos,
+                unsigned char* buf,
+                size_t size,
+                boost::system::error_code& ec)
+            {
+                std::fstream::seekg(pos);
+                std::fstream::read((char*)buf, size);
+                size_ = std::fstream::tellg() - (std::streampos)pos;
+                if (size != size_)
+                {
+                    ec = boost::system::errc::make_error_code(
+                        boost::system::errc::io_error);
+                }
+            }
+            void
+            write(
+                size_t pos,
+                unsigned char* buf,
+                size_t size,
+                boost::system::error_code& ec)
+            {
+                std::fstream::seekp(pos);
+                std::fstream::write((char*)buf, size);
+                size_ = std::fstream::tellp() - (std::streampos)pos;
+                if (size != size_)
+                {
+                    ec = boost::system::errc::make_error_code(
+                        boost::system::errc::io_error);
+                }
+            }
+            size_t
+            size(boost::system::error_code& ec)
+            {
+                return size_;
+            }
+        };
+
+        stream_ keyfile(keypath.string());
+        stream_ datfile(datpath.string());
+
+        if (!keyfile || !datfile)
+            return false;
+
+        nudb::detail::key_file_header keyhead;
+        nudb::detail::dat_file_header dathead;
+        boost::system::error_code ec;
+
+        nudb::detail::read(keyfile, keyhead, ec);
+        if (ec)
+            return false;
+        nudb::detail::read(datfile, dathead, ec);
+        if (ec)
+            return false;
+
+        keyhead.uid = dathead.uid = digest(0);
+
+        nudb::detail::write(keyfile, keyhead, ec);
+        if (ec)
+            return false;
+        nudb::detail::write(datfile, dathead, ec);
+        if (ec)
+            return false;
+    }
+
+    backend_->open(
+        false,
+        digest(2) | 0x5348524400000000ll, /* appType */
```

Comment:
Maybe a small explanation in the comment to this magic number? ("SHRD" in ASCII)

Reply:
Done